routers/web/repo/fork_to_edit.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// CheckPushEditBranch chesk if pushing to the branch in the edit repository is possible,
+// and if not renders an error and returns false.
+func canPushToEditRepository(ctx *context.Context, editRepo *repo_model.Repository, branchName, commitChoice string, tpl templates.TplName, form any) bool {
+	// When pushing to a fork or chosing to commit to a new branch, it should not exist yet
+	if editRepo.ID != ctx.Repo.Repository.ID || commitChoice == frmCommitChoiceNewBranch {
+		if exist, err := git_model.IsBranchExist(ctx, editRepo.ID, branchName); err == nil && exist {
+			ctx.Data["Err_NewBranchName"] = true
+			ctx.RenderWithErr(ctx.Tr("repo.editor.branch_already_exists", branchName), tpl, form)
+			return false
+		}
+	}
+
+	// Check for protected branch
+	canCommitToBranch, err := context.CanCommitToBranch(ctx, ctx.Doer, editRepo, branchName)
+	if err != nil {
+		log.Error("CanCommitToBranch: %v", err)
+	}
+	if !canCommitToBranch.CanCommitToBranch {
+		ctx.Data["Err_NewBranchName"] = true
+		ctx.RenderWithErr(ctx.Tr("repo.editor.cannot_commit_to_protected_branch", branchName), tpl, form)
+		return false
+	}
+
+	return true
+}
+
+// pushToEditRepositoryOrError pushes the branch that editing will be applied on top of
+// to the user fork, if needed. On failure, it displays and returns an error. The
+// branch name to be used for editing is returned.
+func pushToEditRepositoryOrError(ctx *context.Context, editRepo *repo_model.Repository, branchName string, tpl templates.TplName, form any) (string, error) {
+	// If editing the same repository, no need to push anything
+	if editRepo.ID == ctx.Repo.Repository.ID {
+		return ctx.Repo.BranchName, nil
+	}
+
+	// If editing a user fork, first push the branch to that repository
+	baseRepo := ctx.Repo.Repository
+	baseBranchName := ctx.Repo.BranchName
+
+	log.Trace("pushBranchToUserRepo: pushing branch to user repo for editing: %s:%s %s:%s", baseRepo.FullName(), baseBranchName, editRepo.FullName(), branchName)
+
+	if err := git.Push(ctx, baseRepo.RepoPath(), git.PushOptions{
+		Remote: editRepo.RepoPath(),
+		Branch: baseBranchName + ":" + branchName,
+		Env:    repo_module.PushingEnvironment(ctx.Doer, editRepo),
+	}); err != nil {
+		ctx.RenderWithErr(ctx.Tr("repo.editor.fork_failed_to_push_branch", branchName), tpl, form)
+		return "", err
+	}
+
+	return branchName, nil
+}
+
+// updateEditRepositoryIsEmpty updates the the edit repository to mark it as no longer empty
+func updateEditRepositoryIsEmpty(ctx *context.Context, editRepo *repo_model.Repository) {
+	if !editRepo.IsEmpty {
+		return
+	}
+
+	editGitRepo, err := gitrepo.OpenRepository(git.DefaultContext, editRepo)
+	if err != nil {
+		log.Error("gitrepo.OpenRepository: %v", err)
+		return
+	}
+	if editGitRepo == nil {
+		return
+	}
+
+	if isEmpty, err := editGitRepo.IsEmpty(); err == nil && !isEmpty {
+		_ = repo_model.UpdateRepositoryCols(ctx, &repo_model.Repository{ID: editRepo.ID, IsEmpty: false}, "is_empty")
+	}
+	editGitRepo.Close()
```

Comment:
It should use `defer`, or just use RequestContext, then no need to close.

Reply:
I added defer.

I'm unsure about what RequestContext means here. If it's `ctx.Repo.GitRepo`, that's a different repository when editing a fork.

Reply:
After reading context logic, I think it could be like this?

https://github.com/go-gitea/gitea/pull/34240/commits/e78c6c463f3491b8401fcdf49f82a02fcd2eb041

Because when this function is called, we are sure that the repo is never empty.

Reply:
Ah yeah, even better.